// angular-material-rail-drawer/esm5/lib/animations.settings.js
import { group, query, animateChild, animate, style } from '@angular/animations';
import { miniConfig } from './default.config';
export function sidebarClose(minWidth) {
    if (minWidth === void 0) { minWidth = miniConfig.defaultMinWidth; }
    return style({
        width: minWidth,
        visibility: 'visible',
        transform: 'none',
        overflow: 'hidden'
    });
}
export function sidebarOpen(maxWidth) {
    if (maxWidth === void 0) { maxWidth = miniConfig.defaultMaxWidth; }
    return style({
        width: maxWidth,
        visibility: 'visible',
    });
}
export function sidebarAnimationOpenGroup(animationDuration, maxWidth) {
    if (animationDuration === void 0) { animationDuration = miniConfig.defaultDuration; }
    if (maxWidth === void 0) { maxWidth = miniConfig.defaultMaxWidth; }
    return group([
        query('@iconAnimation', animateChild(), { optional: true }),
        query('@labelAnimation', animateChild(), { optional: true }),
        animate(animationDuration + " ease-in-out", sidebarOpen(maxWidth))
    ]);
}
export function sidebarAnimationCloseGroup(animationDuration, minWidth) {
    if (animationDuration === void 0) { animationDuration = miniConfig.defaultDuration; }
    if (minWidth === void 0) { minWidth = miniConfig.defaultMinWidth; }
    return group([
        query('@iconAnimation', animateChild(), { optional: true }),
        query('@labelAnimation', animateChild(), { optional: true }),
        animate(animationDuration + " ease-in-out", sidebarClose(minWidth))
    ]);
}
