// fashion/index.js
var Fashion = require('./src/export/Base.js');
var CssVariableManager = require('./src/export/css/CssVariableManager.js'),
    css = new CssVariableManager();

Fashion.apply(Fashion, {
    css: css,
    CssExport: CssVariableManager,
    ValueParser: require('./src/export/parse/ValueParser.js'),
    functions: {
        Color: require('./src/functions/Color.js'),
        Gradients: require('./src/functions/Gradients.js'),
        HSL: require('./src/functions/HSL.js'),
        Introspection: require('./src/functions/Introspection.js'),
        List: require('./src/functions/List.js'),
        Misc: require('./src/functions/Misc.js'),
        Numeric: require('./src/functions/Numeric.js'),
        Opacity: require('./src/functions/Opacity.js'),
        RGB: require('./src/functions/RGB.js'),
        Selectors: require('./src/functions/Selectors.js'),
        Text: require('./src/functions/Text.js'),
        Util: require('./src/functions/Util.js')
    },
    parse: {
        ast: {
            BaseNode: require('./src/parse/ast/BaseNode.js'),
            Nodes: require('./src/parse/ast/Nodes.js')
        },
        Scanner: require('./src/parse/Scanner.js'),
        Tokenizer: require('./src/parse/Tokenizer.js'),
        Parser: require('./src/parse/Parser.js')
    },
    processors: {
        CssVariables: require('./src/processors/CssVariables.js'),
        NameRegistrations: require('./src/processors/NameRegistrations.js'),
        DataInline: require('./src/processors/DataInline.js')
    },
    BaseSelector: require('./src/type/selectors/BaseSelector.js'),
    BaseSelectorList: require('./src/type/selectors/BaseSelectorList.js'),
    CompoundSelector: require('./src/type/selectors/CompoundSelector.js'),
    MultiPartSelector: require('./src/type/selectors/MultiPartSelector.js'),
    SelectorList: require('./src/type/selectors/SelectorList.js'),
    SelectorPart: require('./src/type/selectors/SelectorPart.js'),
    SelectorProperty: require('./src/type/selectors/SelectorProperty.js'),
    Type: require('./src/export/type/Type.js'),
    Bool: require('./src/export/type/Bool.js'),
    Literal: require('./src/export/type/Literal.js'),
    ParentheticalExpression: require('./src/export/type/ParentheticalExpression.js'),
    Text: require('./src/export/type/Text.js'),
    Numeric: require('./src/export/type/Numeric.js'),
    List: require('./src/export/type/List.js'),
    Map: require('./src/export/type/Map.js'),
    Color: require('./src/export/type/Color.js'),
    ColorRGBA: require('./src/export/type/ColorRGBA.js'),
    ColorHSLA: require('./src/export/type/ColorHSLA.js'),
    ColorStop: require('./src/export/type/ColorStop.js'),
    FunctionCall: require('./src/export/type/FunctionCall.js'),
    Declaration: require('./src/type/Declaration.js'),
    LinearGradient: require('./src/export/type/LinearGradient.js'),
    RadialGradient: require('./src/export/type/RadialGradient.js'),
    Ruleset: require('./src/type/Ruleset.js'),
    Statics: require('./src/export/type/Statics.js'),
    SourceBuilder: require('./src/type/SourceBuilder.js'),
    Types: require('./src/type/Types.js'),
    TypeVisitor: require('./src/type/TypeVisitor.js'),
    Builder: require('./src/Builder.js'),
    CompressedOutput: require('./src/CompressedOutput.js'),
    Context: require('./src/Context.js'),
    CSS: require('./src/CSS.js'),
    CssPostprocessor: require('./src/CssPostprocessor.js'),
    Env: require('./src/Env.js'),
    ExtendProcessor: require('./src/ExtendProcessor.js'),
    Output: require('./src/export/Output.js'),
    PlaceholderProcessor: require('./src/PlaceholderProcessor.js'),
    Preprocessor: require('./src/Preprocessor.js'),
    Ready: require('./src/Ready.js'),
    Runtime: require('./src/Runtime.js'),
    SassFile: require('./src/SassFile.js'),
    Transpiler: require('./src/Transpiler.js'),
    Visitor: require('./src/Visitor.js')

});

module.exports = Fashion;
