// fashion/src/type/Declaration.js
"use strict";

var Fashion = require('../export/Base.js');
var Type = require('./../export/type/Type.js');

class Declaration extends Type {
    constructor(cfg) {
        super();
        if (cfg) {
            Fashion.apply(this, cfg);
        }
    }

    doVisit(visitor) {
        visitor.declaration(this);
    }

    descend(visitor) {
        visitor.visit(this.value);
    }

    copy() {
        return new Declaration({
            property: this.property,
            important: this.important,
            sourceInfo: this.sourceInfo,
            value: this.value.clone()
        });
    }

    clone(match, replace) {
        if (replace && this.matches(match)) {
            return replace.clone()
        }
        return new Declaration({
            property: this.property,
            important: this.important,
            sourceInfo: this.sourceInfo,
            ruleset: this.ruleset,
            value: this.value.clone(match, replace)
        });
    }
}

Fashion.apply(Declaration.prototype, {
    type: 'declaration',
    $isFashionDeclaration: true,
    $canUnbox: false,
    property: null,
    value: null,
    important: null,
    sourceInfo: null,
    ruleset: null
});

module.exports = Declaration;
