// highcharts-angular/bundles/highcharts-angular.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('highcharts-angular', ['exports', '@angular/core'], factory) :
    (factory((global['highcharts-angular'] = {}),global.ng.core));
}(this, (function (exports,core) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var HighchartsChartComponent = /** @class */ (function () {
        function HighchartsChartComponent(el, _zone // #75
        ) {
            this.el = el;
            this._zone = _zone;
            this.updateChange = new core.EventEmitter(true);
            this.chartInstance = new core.EventEmitter();
        }
        Object.defineProperty(HighchartsChartComponent.prototype, "options", {
            set: /**
             * @param {?} val
             * @return {?}
             */ function (val) {
                this.optionsValue = val;
                this.wrappedUpdateOrCreateChart();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(HighchartsChartComponent.prototype, "update", {
            set: /**
             * @param {?} val
             * @return {?}
             */ function (val) {
                if (val) {
                    this.wrappedUpdateOrCreateChart();
                    this.updateChange.emit(false); // clear the flag after update
                }
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        HighchartsChartComponent.prototype.wrappedUpdateOrCreateChart = /**
         * @return {?}
         */
            function () {
                var _this = this;
                // #75
                if (this.runOutsideAngular) {
                    this._zone.runOutsideAngular(function () {
                        _this.updateOrCreateChart();
                    });
                }
                else {
                    this.updateOrCreateChart();
                }
            };
        /**
         * @return {?}
         */
        HighchartsChartComponent.prototype.updateOrCreateChart = /**
         * @return {?}
         */
            function () {
                if (this.chart && this.chart.update) {
                    this.chart.update(this.optionsValue, true, this.oneToOne || false);
                }
                else {
                    this.chart = ( /** @type {?} */(this.Highcharts))[this.constructorType || 'chart'](this.el.nativeElement, this.optionsValue, this.callbackFunction || null);
                    // emit chart instance on init
                    this.chartInstance.emit(this.chart);
                }
            };
        /**
         * @return {?}
         */
        HighchartsChartComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                // #44
                if (this.chart) { // #56
                    // #56
                    this.chart.destroy();
                    this.chart = null;
                }
            };
        HighchartsChartComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'highcharts-chart',
                        template: ''
                    }] }
        ];
        /** @nocollapse */
        HighchartsChartComponent.ctorParameters = function () {
            return [
                { type: core.ElementRef, },
                { type: core.NgZone, },
            ];
        };
        HighchartsChartComponent.propDecorators = {
            "Highcharts": [{ type: core.Input },],
            "constructorType": [{ type: core.Input },],
            "callbackFunction": [{ type: core.Input },],
            "oneToOne": [{ type: core.Input },],
            "runOutsideAngular": [{ type: core.Input },],
            "options": [{ type: core.Input },],
            "update": [{ type: core.Input },],
            "updateChange": [{ type: core.Output },],
            "chartInstance": [{ type: core.Output },],
        };
        return HighchartsChartComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var HighchartsChartModule = /** @class */ (function () {
        function HighchartsChartModule() {
        }
        HighchartsChartModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [HighchartsChartComponent],
                        exports: [HighchartsChartComponent]
                    },] }
        ];
        /** @nocollapse */
        HighchartsChartModule.ctorParameters = function () { return []; };
        return HighchartsChartModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.HighchartsChartModule = HighchartsChartModule;
    exports.HighchartsChartComponent = HighchartsChartComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
