// angular-material-rail-drawer/esm2015/lib/animations.settings.js
import { group, query, animateChild, animate, style } from '@angular/animations';
import { miniConfig } from './default.config';
export function sidebarClose(minWidth = miniConfig.defaultMinWidth) {
    return style({
        width: minWidth,
        visibility: 'visible',
        transform: 'none',
        overflow: 'hidden'
    });
}
export function sidebarOpen(maxWidth = miniConfig.defaultMaxWidth) {
    return style({
        width: maxWidth,
        visibility: 'visible',
    });
}
export function sidebarAnimationOpenGroup(animationDuration = miniConfig.defaultDuration, maxWidth = miniConfig.defaultMaxWidth) {
    return group([
        query('@iconAnimation', animateChild(), { optional: true }),
        query('@labelAnimation', animateChild(), { optional: true }),
        animate(`${animationDuration} ease-in-out`, sidebarOpen(maxWidth))
    ]);
}
export function sidebarAnimationCloseGroup(animationDuration = miniConfig.defaultDuration, minWidth = miniConfig.defaultMinWidth) {
    return group([
        query('@iconAnimation', animateChild(), { optional: true }),
        query('@labelAnimation', animateChild(), { optional: true }),
        animate(`${animationDuration} ease-in-out`, sidebarClose(minWidth))
    ]);
}
