// fashion/src/export/Output.js
/*
 * Copyright (c) 2012-2016. Sencha Inc.
 */

"use strict";

var Fashion = require('./Base.js'),
    Base = Fashion.Base;

class Output extends Base {

    constructor() {
        super();
        this.output = '';
    }

    space() {
        this.add(' ');
    }

    add(text) {
        this.output += text;
    }

    addComment(text) {
        this.output += text;
    }

    indent() {
        this.indentation += this.indentstr;
    }

    unindent() {
        this.indentation = this.indentation.substr(this.indentstr.length);
    }

    addln(ln) {
        this.output += '\n' + this.indentation + (ln || '');
    }

    addCommentLn(ln) {
        if (ln && ln.indexOf('//') === 0) {
            return;
        }
        this.addln(ln)
    }

    get() {
        return this.output;
    }

    indentln(ln) {
        this.addln(ln);
        this.indent();
    }

    unindentln(ln) {
        this.unindent();
        this.addln(ln);
    }

    reset() {
        this.indentation = '';
        this.output = '';
    }
}

Fashion.apply(Output.prototype, {
    indentation: '',
    output: '',
    isCompressed: false,
    indentstr: '    ',
    splitThreshold: 1000000,
    selectorCount: 0
});

module.exports = Output;
