// fashion/src/type/selectors/BaseSelectorList.js
"use strict";

var Fashion = require('../../export/Base.js');
var BaseSelector = require('./BaseSelector.js');

class BaseSelectorList extends BaseSelector {
    constructor(items, separator) {
        super();
        this.items = items;
        this.separator = separator;
    }

    toString() {
        return this.items.join(this.separator);
    }

    cloneItems(match, replace) {
        var cloned = [];
        for (var i = 0; i < this.items.length; i++) {
            cloned.push(this.items[i].clone(match, replace));
        }
        return cloned;
    }

    unquote() {
        var items = [];
        for (var i = 0; i < this.items.length; i++) {
            if (this.items[i]) {
                items.push(this.items[i].unquote());
            }
        }
        return new List(items, this.separator);
    }

    _getHash() {
        var items = [],
            tItems = this.items,
            item;

        for (var i = 0; i < tItems.length; i++) {
            item = tItems[i];
            items.push((item && item.getHash()) || '');
        }

        items = items.sort();
        return items.join(this.separator);
    }

    setItems (items) {
        this.items = items;
        this._hash = null;
    }
}

Fashion.apply(BaseSelectorList.prototype, {
    separator: null,
    $isFashionBaseSelectorList: true
});

Fashion.mixin(BaseSelectorList, Fashion.BaseSet);

module.exports = BaseSelectorList;
