// fashion/src/type/selectors/SelectorList.js
"use strict";

var Fashion = require('../../export/Base.js');
var BaseSelectorList = require('./BaseSelectorList.js');

class SelectorList extends BaseSelectorList {
    constructor(list) {
        super(list, ',');
    }

    doVisit(visitor) {
        visitor.selectorlist(this);
    }

    descend(visitor) {
        for (var i = 0; i < this.items.length; i++) {
            var item = this.items[i];
            item && visitor.visit(item);
        }
    }

    clone(match, replace) {
        var cloned = new SelectorList(this.cloneItems());
        if (this.parent) {
            cloned.setParent(this.parent.clone(match, replace));
        }
        return cloned;
    }

    applyInterpolations() {
        if (!this.interpolated) {
            this.interpolated = true;
            var interpolated = [],
                selectors = this.items,
                selector, str, items;

            for (var s = 0; s < selectors.length; s++) {
                selector = selectors[s];
                str = selector.toString();
                if (str.indexOf(',') === -1) {
                    interpolated.push(selector);
                } else {
                    items = str.split(',');
                    interpolated.push.apply(interpolated, items);
                }
            }
            this.items = interpolated;
        }
    }
    
    flatten () {
        var items = this.items,
            len = items.length,
            newItems = [],
            item;
        for (var i = 0; i < len; i++) {
            item = items[i];
            while (item && item.visitTarget) {
                item = item.visitTarget;
            }
            if (item.$isFashionSelectorList) {
                newItems.push.apply(newItems, item.flatten());
            }
            else {
                if (item.$isFashionMultiPartSelector || item.$isFashionCompoundSelector) {
                    item.flatten();
                }
                newItems.push(item);
            }
        }
        this.items = newItems;
        return newItems;
    }
}

Fashion.apply(SelectorList.prototype, {
    $isFashionSelectorList: true,
    type: 'selectorlist',
    ruleset: null,
    interpolated: null
});

module.exports = SelectorList;
