// @angular/compiler-cli/ngcc/src/execution/cluster/worker.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/ngcc/src/execution/cluster/worker", ["require", "exports", "cluster", "@angular/compiler-cli/ngcc/src/execution/tasks/utils", "@angular/compiler-cli/ngcc/src/execution/cluster/utils"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /// <reference types="node" />
    var cluster = require("cluster");
    var utils_1 = require("@angular/compiler-cli/ngcc/src/execution/tasks/utils");
    var utils_2 = require("@angular/compiler-cli/ngcc/src/execution/cluster/utils");
    /**
     * A cluster worker is responsible for processing one task (i.e. one format property for a specific
     * entry-point) at a time and reporting results back to the cluster master.
     */
    var ClusterWorker = /** @class */ (function () {
        function ClusterWorker(logger, createCompileFn) {
            this.logger = logger;
            if (cluster.isMaster) {
                throw new Error('Tried to instantiate `ClusterWorker` on the master process.');
            }
            this.compile = createCompileFn(function (_task, outcome, message) {
                return utils_2.sendMessageToMaster({ type: 'task-completed', outcome: outcome, message: message });
            });
        }
        ClusterWorker.prototype.run = function () {
            var _this = this;
            // Listen for `ProcessTaskMessage`s and process tasks.
            cluster.worker.on('message', function (msg) {
                try {
                    switch (msg.type) {
                        case 'process-task':
                            _this.logger.debug("[Worker #" + cluster.worker.id + "] Processing task: " + utils_1.stringifyTask(msg.task));
                            return _this.compile(msg.task);
                        default:
                            throw new Error("[Worker #" + cluster.worker.id + "] Invalid message received: " + JSON.stringify(msg));
                    }
                }
                catch (err) {
                    utils_2.sendMessageToMaster({
                        type: 'error',
                        error: (err instanceof Error) ? (err.stack || err.message) : err,
                    });
                }
            });
            // Return a promise that is never resolved.
            return new Promise(function () { return undefined; });
        };
        return ClusterWorker;
    }());
    exports.ClusterWorker = ClusterWorker;
});
